Gameflow is put together

Katerina-codes opened the pull request:
Progress on Rock, Paper, Scissors so far. 

**Features included**

- Scoring the game for each player's move.
- Asking a player for a move, getting the move.
- Overall game flow.
- Added  Greek language option.
- Removed Language parameter on Commandline class.


1.  I need to fix the 'draw' announcement.
2. No validation throughout program.
3. Need to add game mode options for Human vs Human and Human vs Computer.

@MollieS @maikon @jsuchy

Katerina-codes committed:
Announce winner

Katerina-codes committed:
asksUserToEnterMove test uses new constructor

- Removed old constructor.

Katerina-codes committed:
Removed duplication with setUp method before tests

Katerina-codes committed:
CommandlineUi implements Ui interface

- Created Ui interface and FakeCommandlineUi which also implements
Ui interface.
- Refactored display methods so they return void.
- Corrected assertion and test setup.
- Added Game class and tests to check methods are being called.

Katerina-codes committed:
Added call to getsMove method in runGame

- Renamed ui variable to inputOutput for consistency.
- Add scoreGame & announceWinner methods to newGame
- Added another runGame method with doubles for all methods in runGame.

Katerina-codes committed:
Added conversion methods

- Added announceWinnerTwo method.

Katerina-codes committed:
Prepare code for switching runGame with runGameTwo

- Created scoreGameTwo method which uses Result logic only in preparation to remove string values from enums.

Katerina-codes committed:
Added player two move prompt

Katerina-codes committed:
Add gameRunner & move hashmap logic to MoveFactory

Katerina-codes committed:
Added README

Maikon commented on src/main/game/CommandlineUi.java:
```diff
@@ -29,4 +26,43 @@ public String getsMove() {
+        HashMap<String, Moves> convertedMove = new HashMap<>();
+        convertedMove.put("rock", Moves.ROCK);
+        convertedMove.put("paper", Moves.PAPER);
+        convertedMove.put("scissors", Moves.SCISSORS);
```
Consider moving this configuration outside of this method. Perhaps its something that can be done as part of the initialization of this object. Reason for this is because you don't want to create a new map and insert 3 values every time you call `convertMove`. It should be done just once.

Maikon commented on src/main/game/CommandlineUi.java:
```diff
@@ -29,4 +26,43 @@ public String getsMove() {
+        HashMap<Result, String> convertedResult = new HashMap<>();
+        convertedResult.put(Result.PLAYER_ONE_WINS, "Player one wins!");
+        convertedResult.put(Result.PLAYER_TWO_WINS, "Player two wins!");
+        convertedResult.put(Result.DRAW, "It's a draw!");
```
Similar comment to the one above. Consider moving this out of this method and performing the operation once.

Maikon commented on src/main/game/CommandlineUi.java:
```diff
@@ -1,23 +1,20 @@
+    private final PrintStream output;
+    private final BufferedReader input;
 
     public CommandlineUi(PrintStream output, InputStream input) {
```
Be careful of naming conventions: `CommandlineUi` -> `CommandLineUI`

Maikon commented on src/main/game/CommandlineUi.java:
```diff
@@ -29,4 +26,43 @@ public String getsMove() {
         return userMove;
     }
+
+    public Moves getsMoveTwo() {
```
Prefer `get` vs `gets`. This `CommandLineUI.getMove()` reads better than `CommandLineUI.getsMove()`. In the first one you're sending a message to (telling) an object to do something whereas the second one you're describing something that _might_ happen.

Maikon commented on src/main/game/Game.java:
```diff
@@ -0,0 +1,31 @@
+
+    public void runGameTwo() {
+        inputOutput.askForMove();
+        Moves playerOneMove = inputOutput.getsMoveTwo();
```
I would consider adding the prompting of the user for a move inside `getMove()`. It would tidy up the code here and makes it more declarative too vs having procedural steps (i.e do step 1, then step 2 etc)

Maikon commented on src/main/game/GameRunner.java:
```diff
@@ -0,0 +1,11 @@
+        Game game = new Game(inputOutput, rules);
+        game.runGame();
+    }
+}
```
Missing new line.

Maikon commented on src/main/game/MoveFactory.java:
```diff
@@ -0,0 +1,22 @@
+        Move scissorsMove = new Scissors();
+        moves.put(ROCK, rockMove);
+        moves.put(PAPER, paperMove);
+        moves.put(SCISSORS, scissorsMove);
```
Similar comment with the `CommandLineUI` comments. Move this logic in a place where it only gets executed when the object is initialised.

Maikon commented on src/main/game/Result.java:
```diff
@@ -1,7 +1,7 @@
 
 public enum Result {
-    DRAW("draw");
+    DRAW("draw"), PLAYER_ONE_WINS(""), PLAYER_TWO_WINS("");
```
Consider doing what you did with `Moves` by removing the strings altogether and do the translation at the boundaries of your system (in, out). Also, `PLAYER_TWO_WINS` is not used.

Maikon commented on src/main/game/Ui.java:
```diff
@@ -0,0 +1,10 @@
+package main.game;
+
+public interface Ui {
```
`Ui` -> `UI`. See _["Java Naming Convention with Acronyms"](https://stackoverflow.com/questions/2236807/java-naming-convention-with-acronyms)_

Maikon commented on src/test/game/CommandlineUiTest.java:
```diff
@@ -9,22 +10,40 @@
+    public void setUp() {
+        output = new ByteArrayOutputStream();
+        input = new ByteArrayInputStream("".getBytes());
+        ui = new CommandlineUi(new PrintStream(output), input);
```
Instead of this code living in the `setUp` method, you could have it in a private helper function that can take in the arguments that tend to change with each test. That would help eliminate duplicated boilerplate code like the one in `getMoveFromUser` test.

Maikon commented on src/main/game/MoveFactory.java:
```diff
@@ -0,0 +1,22 @@
+
+public class MoveFactory {
+
+    private Map moves;
```
Don't forget to add the types of the keys and values.

Maikon commented on src/test/game/GameTest.java:
```diff
@@ -0,0 +1,40 @@
+
+public class GameTest {
+
+    private Map moves;
```
Same comment as above about types.

Maikon commented on src/test/game/RulesTest.java:
```diff
@@ -25,36 +25,43 @@ public void setUp() {
         Rules rules = new Rules(moves);
 
-        assertEquals(DRAW.getResult(), rules.scoreGameTwo(ROCK, ROCK));
+        assertEquals(DRAW.getResult(), rules.scoreGame(ROCK, ROCK));
```
When you apply the refactoring recommended for the `Result` enum this would eliminate the need for `getResult()`. Same would apply for the rest of the tests here.

Maikon reviewed (commented):
Overall good job 👍

Katerina-codes committed:
Initialised Commandline with hashmap

- Removed hashmap from method.

Katerina-codes committed:
Corrected test

Katerina-codes committed:
Player 2 Wins if player 2 move matches winningMove

Katerina-codes committed:
Reformatted code and removed dead code

Katerina-codes committed:
Corrected FakeCommandlineUi

Katerina-codes committed:
Added test for runGameTwo method

- Removed unnecessary code from the FakeCommandlineUi

Katerina-codes committed:
Replace methods that use strings with new methods

- Removed strings from enums.
- Removed all old methods that were no longer used.
- Corrected tests.
- Made scoreGameTwo method return a result rather than a winning move.

Katerina-codes committed:
Renamed methods for better clarity

Katerina-codes committed:
Added Language ui and English class

- Added CommandlineUi double with new constructor that has knowledge of
Language.
- Added askForMoveTwo method on CommandlineUi.
- Extracted both hashmaps into methods of their own and initialised
CommandlineUi with both methods.

Katerina-codes committed:
Added curly braces to If Statement for consistency

Katerina-codes committed:
English implements language interface

Katerina-codes committed:
Test uses new constructor and askForMoveTwo method

- Test uses new Language interface logic. Method now more abstract.

Katerina-codes committed:
Use announceWinnerTwo method

- Corrected FakeCommandlineUi.
- Changed CommandlineUi test so it uses announceWinnerTwo method.

Katerina-codes committed:
Add calls to new methods in Game class and tests

- Removing dependency on hard coded string concretions.
- Removed methods that are no longer used.
- Swapped the new CommandlineUi constructor logic into the GameRunner.
- Removed old constructor with Language logic from CommandlineUi and
tests.

Katerina-codes committed:
Move duplicated logic to setUp method

Katerina-codes committed:
Prompt user to language choice

Katerina-codes committed:
Get user language choice

Katerina-codes committed:
Corrected indentation

Katerina-codes committed:
Displays english is user enters '1'

Katerina-codes committed:
Displays greek if user enters '2'

Katerina-codes committed:
Sets language to english

Katerina-codes committed:
Sets language to greek

Katerina-codes committed:
Add new constructor without Language parameter

- Corrected askForLanguage.
- Corrected test so it's consistent.

Katerina-codes committed:
Remove old constructor

- English is set to default language.
- setUserLanguage method now called to provide all language
functionality.

Katerina-codes committed:
Remove old constructor

- English is set to default language.
- setUserLanguage method now called to provide all language
functionality.

Katerina-codes committed:
Added real translation to Greek language

- Corrected indentation.

Katerina-codes committed:
Translates greek moves to english

Katerina-codes committed:
Fixed merge conflicts

Katerina-codes committed:
Fixed merge conflicts

Katerina-codes committed:
Inline code. Return variable not assignment

Katerina-codes committed:
Correct greek translations and indentation

Katerina-codes committed:
Organise files by namespace

- Removed dead code.

Katerina-codes committed:
Remove unused imports

Katerina-codes committed:
Removed unused code

Katerina-codes committed:
Use helper method to remove duplication

Katerina-codes committed:
Renamed ui variable

Katerina-codes committed:
Add all tests for English

Katerina-codes committed:
Removed unused code

Katerina-codes committed:
Changed tests to test behaviour more accurately

- Removed unnecessary method and got method to return void.

Katerina-codes committed:
Fixed merge conflicts

Katerina-codes committed:
Added tests to Greek class. Added GreekTest class
